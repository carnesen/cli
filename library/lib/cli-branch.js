"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliBranch = void 0;
const constants_1 = require("./constants");
/**
 * A factory function for creating command branches
 * @remarks
 * @param options
 * @returns Returns the newly-created branch object
 * @example
 * ```typescript
 * const cloudBranch = CliBranch()
 * ```
 */
function CliBranch(options) {
    return {
        ...options,
        commandType: constants_1.CLI_BRANCH,
    };
}
exports.CliBranch = CliBranch;
//# sourceMappingURL=cli-branch.js.map