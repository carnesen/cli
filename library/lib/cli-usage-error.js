"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliUsageError = exports.CLI_USAGE_ERROR = void 0;
exports.CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';
/**
 * A custom error with code {@linkcode CLI_USAGE_ERROR}
 * @remarks
 * {@linkcode runCliAndExit} catches this class of error and prints command-line usage
 * (i.e. "Usage: cloud users lists ...") for the current command. If the `message`
 * property is provided, `runCliAndExit` will normal `Error`. See also print `Error: <your
 * message>`. See also {@linkcode CliTerseError}.
 * @example
 * ```typescript
 * throw new CliUsageError("Expected one or more values")
 * ```
 */
class CliUsageError extends Error {
    /**
     *
     * @param message An `Error` message
     * @param commandStack Used internally for argument processing
     */
    constructor(message, commandStack) {
        super(message);
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandStack", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.code = exports.CLI_USAGE_ERROR;
        this.commandStack = commandStack;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.CliUsageError = CliUsageError;
//# sourceMappingURL=cli-usage-error.js.map