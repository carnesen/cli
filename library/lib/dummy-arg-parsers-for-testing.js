"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dummyRequiredArgParser = exports.dummyArgParser = exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY = exports.DUMMY_ARG_PARSER_THROW = exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = void 0;
exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = 'undefined was passed';
exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = 'undefined was passed';
exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = 'thrown intentionally';
exports.DUMMY_ARG_PARSER_THROW = 'throw';
exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY = 'throw-non-truthy';
exports.dummyArgParser = {
    placeholder: '',
    parse(args) {
        if (typeof args === 'undefined') {
            return exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED;
        }
        if (args.length === 0) {
            return exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED;
        }
        if (args[0] === exports.DUMMY_ARG_PARSER_THROW) {
            throw new Error(exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY);
        }
        if (args[0] === exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY) {
            // eslint-disable-next-line no-throw-literal
            throw '';
        }
        return args[0];
    },
};
exports.dummyRequiredArgParser = {
    placeholder: '<foo>',
    required: true,
    parse: exports.dummyArgParser.parse,
};
//# sourceMappingURL=dummy-arg-parsers-for-testing.js.map