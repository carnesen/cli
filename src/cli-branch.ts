/* eslint-disable no-redeclare */
import { CliBranch, ExcludeCommandType } from './types';
import { CLI_BRANCH } from './constants';

type Config = ExcludeCommandType<CliBranch>;

/**
 * A factory function for creating command branches
 * @remarks
 * @param config
 * @returns Returns the newly-created `CliBranch` object
 */
export function CliBranch(config: Config): CliBranch {
  const branch: CliBranch = {
    ...config,
    commandType: CLI_BRANCH,
  };
  return branch;
}
