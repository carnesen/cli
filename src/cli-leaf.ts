/* eslint-disable no-redeclare */
import {
  AnyNamedArgParsers,
  CliLeaf,
  ExcludeCommandType,
  AnyArgParser,
  CliArgParser,
} from './types';
import { CLI_LEAF } from './constants';

export function CliLeaf<
  TPositional extends AnyArgParser = CliArgParser<undefined, false>,
  TNamed extends AnyNamedArgParsers = any,
  TEscaped extends AnyArgParser = CliArgParser<undefined, false>
>(
  config: ExcludeCommandType<CliLeaf<TPositional, TNamed, TEscaped>>,
): CliLeaf<TPositional, TNamed, TEscaped> {
  const cliLeaf: CliLeaf<TPositional, TNamed, TEscaped> = {
    ...config,
    commandType: CLI_LEAF,
  };
  return cliLeaf;
}
