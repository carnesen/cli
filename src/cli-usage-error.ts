import { CommandStack } from './types';

export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

export class CliUsageError extends Error {
  public readonly code: typeof CLI_USAGE_ERROR;

  public commandStack?: CommandStack;

  constructor(message?: string, commandStack?: CommandStack) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    this.commandStack = commandStack;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
