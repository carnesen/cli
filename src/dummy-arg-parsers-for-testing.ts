import { CliArgParser } from './types';

export const DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = 'undefined was passed';
export const DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = 'undefined was passed';
export const DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = 'thrown intentionally';
export const DUMMY_ARG_PARSER_THROW = 'throw';
export const DUMMY_ARG_PARSER_THROW_NON_TRUTHY = 'throw-non-truthy';

export const dummyArgParser: CliArgParser<string, false> = {
  placeholder: '',
  getValue(args) {
    if (typeof args === 'undefined') {
      return DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED;
    }
    if (args.length === 0) {
      return DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED;
    }
    if (args[0] === DUMMY_ARG_PARSER_THROW) {
      throw new Error(DUMMY_ARG_PARSER_THROWN_INTENTIONALLY);
    }
    if (args[0] === DUMMY_ARG_PARSER_THROW_NON_TRUTHY) {
      // eslint-disable-next-line no-throw-literal
      throw '';
    }
    return args[0];
  },
};

export const dummyRequiredArgParser: CliArgParser<string, true> = {
  placeholder: '<foo>',
  required: true,
  getValue: dummyArgParser.getValue,
};
