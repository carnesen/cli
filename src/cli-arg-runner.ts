import { AnyCommand } from './types';
import { accumulateCommandStack } from './accumulate-command-stack';
import { accumulateArgs } from './accumulate-args';
import { accumulateNamedValues } from './accumulate-named-values';
import { CliUsageError } from './cli-usage-error';
import { CliTerseError } from './cli-terse-error';
import { CLI_BRANCH } from './constants';
import { findVersion } from './find-version';
import { callParse } from './call-parse';
import { LastCommand } from './last-command';

export type CliArgRunner = (...args: string[]) => Promise<any>;

export type CliEnhancer = (argRunner: CliArgRunner) => CliArgRunner;

export function CliArgRunner(
  rootCommand: AnyCommand,
  options: Partial<{ enhancer: CliEnhancer }> = {},
): CliArgRunner {
  const { enhancer } = options;

  if (enhancer) {
    return enhancer(cliArgRunner);
  }

  return cliArgRunner;

  async function cliArgRunner(...args: string[]) {
    if (['-v', '--version'].includes(args[0])) {
      const version = await findVersion();
      if (!version) {
        throw new CliTerseError('Failed to find a "version" string');
      }
      return version;
    }

    const {
      foundHelp,
      commandNamesAndPositionalArgs,
      namedArgs,
      escapedArgs,
    } = accumulateArgs(args);

    const restCommandNamesAndPositionalArgs = accumulateCommandStack(
      rootCommand,
      commandNamesAndPositionalArgs,
    );

    if (foundHelp) {
      // E.g.:
      //   cli branch0 branch1 --help
      // Same as:
      //   cli --help branch0 branch1
      throw new CliUsageError();
    }

    const lastCommand = LastCommand(rootCommand);

    if (lastCommand.commandType === CLI_BRANCH) {
      if (restCommandNamesAndPositionalArgs[0]) {
        // E.g. cli branch0 branch1 bad-command-name
        throw new CliUsageError(`Bad command "${restCommandNamesAndPositionalArgs[0]}"`);
      }
      // E.g. cli branch0 branch1
      throw new CliUsageError();
    }

    let argsValue: any;
    if (lastCommand.positionalArgParser) {
      // Note that for named and escaped args, we distinguish between
      // `undefined` and `[]`. For example, "cli" gives an escaped args
      // `undefined` whereas "cli --" gives an escaped args `[]`. For the
      // "positionalArgs", however, there is no such distinction. By convention,
      // we elect here to pass in `undefined` rather than an empty array when no
      // positional arguments are passed.
      const positionalArgs =
        restCommandNamesAndPositionalArgs.length > 0
          ? restCommandNamesAndPositionalArgs
          : undefined;
      argsValue = await callParse(lastCommand.positionalArgParser, positionalArgs);
    } else if (restCommandNamesAndPositionalArgs.length > 0) {
      throw new CliUsageError(
        `Unexpected argument "${restCommandNamesAndPositionalArgs[0]}" : Command "${lastCommand.name}" does not accept positional arguments`,
      );
    }

    const namedValues = await accumulateNamedValues(
      lastCommand.namedArgParsers || {},
      namedArgs,
    );

    let escapedValue: any;
    if (lastCommand.escapedArgParser) {
      escapedValue = await callParse(lastCommand.escapedArgParser, escapedArgs, '--');
    } else if (escapedArgs) {
      throw new CliUsageError(
        `Command "${lastCommand.name}" does not allow "--" as an argument`,
      );
    }

    const result = await lastCommand.action(argsValue, namedValues, escapedValue);
    return result;
  }
}
