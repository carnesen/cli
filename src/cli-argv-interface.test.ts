import { runAndCatch } from '@carnesen/run-and-catch';
import { CliBranch } from './cli-branch';
import { CliLeaf } from './cli-leaf';
import { dummyArgParser } from './dummy-arg-parsers-for-testing';
import { CliArgRunner, CliEnhancer } from './cli-arg-runner';
import { findVersion } from './find-version';
import { CLI_USAGE_ERROR } from './cli-usage-error';

const leafWithNamedArgParsers = CliLeaf({
  name: 'leaf-with-named-args',
  namedArgParsers: {
    foo: dummyArgParser,
  },
  action(...args) {
    return args;
  },
});

const leafWithPositionalArgParser = CliLeaf({
  name: 'leaf-with-positional-args',
  positionalArgParser: dummyArgParser,
  action(...args) {
    return args;
  },
});

const leafWithEscapedArgParser = CliLeaf({
  name: 'leaf-with-escaped-arg-parser',
  escapedArgParser: dummyArgParser,
  action(...args) {
    return args;
  },
});

const root = CliBranch({
  name: 'cli',
  subcommands: [
    leafWithPositionalArgParser,
    leafWithNamedArgParsers,
    leafWithEscapedArgParser,
  ],
});

const cliArgRunner = CliArgRunner(root);

describe(CliArgRunner.name, () => {
  it('calls the enhancer if provided', async () => {
    const spy = jest.fn();
    const enhancer: CliEnhancer = (innerArgRunner) => async (...args: string[]) => {
      spy(...args);
      await innerArgRunner(...args);
    };
    const enhancedArgRunner = CliArgRunner(leafWithPositionalArgParser, {
      enhancer,
    });
    await enhancedArgRunner('foo', 'bar');
    expect(spy.mock.calls).toEqual([['foo', 'bar']]);
  });

  it('returns version string from package.json if "-v" or "--version" is passed', async () => {
    const version = await findVersion();
    expect(await cliArgRunner('-v')).toBe(version);
    expect(await cliArgRunner('--version')).toBe(version);
  });

  it('throws USAGE error with empty message if --help is passed', async () => {
    const exception = await runAndCatch(cliArgRunner, '--help');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error with empty message if last command is a branch and no additional args is present', async () => {
    const exception = await runAndCatch(cliArgRunner);
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error "bad command" if last command is a branch and additional args is present', async () => {
    const exception = await runAndCatch(cliArgRunner, 'oops');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(/bad command/i);
    expect(exception.message).toMatch('"oops"');
    expect(exception.message).toMatchSnapshot();
  });

  it('throws USAGE error "positional arguments" if last command is a leaf without positionalArgParser property and additional args is present', async () => {
    const exception = await runAndCatch(
      cliArgRunner,
      leafWithNamedArgParsers.name,
      'oops',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch('Unexpected argument "oops"');
    expect(exception.message).toMatch(leafWithNamedArgParsers.name);
    expect(exception.message).toMatch(/positional arguments/i);
    expect(exception.message).toMatchSnapshot();
  });

  it('Passes parsed positional value as first argument of the "action" function', async () => {
    const positionalArgs = ['foo', 'bar'];
    const result = await cliArgRunner(
      leafWithPositionalArgParser.name,
      ...positionalArgs,
    );
    expect(result).toEqual([dummyArgParser.getValue(positionalArgs), {}, undefined]);
  });

  it('Passes parsed named values as second argument of the "action" function', async () => {
    const namedArgs = ['--foo', 'bar'];
    const result = await cliArgRunner(leafWithNamedArgParsers.name, ...namedArgs);
    expect(result).toEqual([
      undefined,
      { foo: dummyArgParser.getValue(['bar']) },
      undefined,
    ]);
  });

  it(`Throws USAGE error 'does not allow "--"' if leaf does not have an "escaped" property`, async () => {
    const exception = await runAndCatch(
      cliArgRunner,
      leafWithPositionalArgParser.name,
      '--',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(leafWithPositionalArgParser.name);
    expect(exception.message).toMatch('does not allow "--"');
  });

  it('Passes parsed escaped value as third argument of the "action" function', async () => {
    const result = await cliArgRunner(leafWithEscapedArgParser.name, '--');
    expect(result).toEqual([
      undefined,
      {},
      leafWithEscapedArgParser.escapedArgParser!.getValue([]),
    ]);
  });
});
