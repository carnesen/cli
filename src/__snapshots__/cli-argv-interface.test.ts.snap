// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CliArgRunner throws USAGE error "bad command" if last command is a branch and additional args is present 1`] = `"Bad command \\"oops\\""`;

exports[`CliArgRunner throws USAGE error "positional arguments" if last command is a leaf without positionalArgParser property and additional args is present 1`] = `"Unexpected argument \\"oops\\" : Command \\"leaf-with-named-args\\" does not accept positional arguments"`;
